File: leetcode_problems/python/lc739_daily_temperatures.py
class Solution:
    def dailyTemperatures(
        self,
        temperatures: list[int],
    ) -> list[int]:
        result = [0 for _ in range(len(temperatures))]
        stack: list[tuple[int, int]] = []
        for index, temp in enumerate(temperatures):
            while stack and stack[-1][1] < temp:
                target, _ = stack.pop()
                result[target] = index - target
            stack.append((index, temp))
        return result


if __name__ == "__main__":
    temperatures = [73, 74, 75, 71, 69, 72, 76, 73]
    print(Solution().dailyTemperatures(temperatures))

File: leetcode_problems/go/lc739_daily_temperatures.go
package main

type DailyTemp struct {
	i int
	v int
}

func dailyTemperatures(temperatures []int) []int {
	stk := make([]DailyTemp, 0)
	res := make([]int, len(temperatures))
	for i, v := range temperatures {
		t := DailyTemp{i, v}
		for len(stk) > 0 {
			p := stk[len(stk)-1]
			if p.v < t.v {
				res[p.i] = t.i - p.i
				stk = stk[:len(stk)-1]
			} else {
				break
			}
		}
		stk = append(stk, t)
	}
	for len(stk) > 0 {
		p := stk[len(stk)-1]
		res[p.i] = 0
		stk = stk[:len(stk)-1]
	}
	return res
}
